### PR TITLE
Resubmit #29747. "Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s"

### PR DESCRIPTION
**Pull request by xush6528**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30093 Resubmit "Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s"**
* #30092 Resubmit "Add RpcAgentTestFixture to extract duplicate code"

https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_rpc` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in https://github.com/pytorch/pytorch/issues/29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

Differential Revision: [D18595578](https://our.internmc.facebook.com/intern/diff/D18595578/)

### CODE REVIEW

**Review by mrshenli — COMMENTED** [review]
Is this exactly the same as the previous PR? And the previous PR was reverted because the PR underneath it was reverted?

**Review by mrshenli — APPROVED** [review]
Synced offline, no change compared to the previous attempt. Stamping to unblock, but please do wait for CI tests.

**Comment by pietern** [comment]
Can you link the original PRs and the revert commits for both resubmits? Spelunking through the source history will be many times easier if this information is included in the commit messages.

**Comment by xush6528** [comment]
Redoing https://github.com/pytorch/pytorch/pull/29747.

**Comment by pietern** [comment]
@xush6528 For future resubmits, can you include both the original PR and the revert commit in the commit summary? Then you can view the Git history locally and trace back your steps.

**Comment by xush6528** [comment]
@pietern You mean the summary section is included into git history. Not the title.

**Comment by pietern** [comment]
@xush6528 Indeed. The title of the PR is use as the title of the commit.

**Pull request closed without merging**
